thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/dataquality/components/DataSlaQualityChecker.java

```diff
@@ -0,0 +1,215 @@
+    this.sla = spec.getSla();
+    this.dataFetcher = dataFetcher;
+  }
+
+  /**
+   * Runs the data sla check for the window on the given metric
+   */
+  private List<MergedAnomalyResultDTO> runSLACheck(MetricEntity me, Interval window) {
+    List<MergedAnomalyResultDTO> anomalies = new ArrayList<>();
+
+    // We want to measure the overall dataset availability (filters can be ignored)
+    long startTime = window.getStart().getMillis();
+    long endTime = window.getEnd().getMillis();
+    MetricSlice metricSlice = MetricSlice.from(me.getId(), startTime, endTime, ArrayListMultimap.<String, String>create());
+    InputData data = this.dataFetcher.fetchData(new InputDataSpec()
+        .withTimeseriesSlices(Collections.singletonList(metricSlice))
+        .withMetricIdsForDataset(Collections.singletonList(me.getId()))
+        .withMetricIds(Collections.singletonList(me.getId())));
+    DatasetConfigDTO datasetConfig = data.getDatasetForMetricId().get(me.getId());
+
+    try {
+      long datasetLastRefreshTime = datasetConfig.getLastRefreshTime();
+      if (datasetLastRefreshTime <= 0) {
+        // no availability event -> assume we have processed data till the current detection start
+        datasetLastRefreshTime = startTime - 1;
+      }
+
+      MetricSlice slice = MetricSlice.from(me.getId(), datasetLastRefreshTime + 1, endTime);
+      if (isMissingData(datasetLastRefreshTime, startTime)) {
+        // Double check with data source as 2 things are possible.
+        // 1. This dataset/source may not support availability events
+        // 2. The data availability event pipeline has some issue.
+
+        DataFrame dataFrame = data.getTimeseries().get(metricSlice);
+        if (dataFrame == null || dataFrame.isEmpty()) {
+          // no data
+          if (hasMissedSLA(datasetLastRefreshTime, endTime)) {
+            anomalies.add(createDataSLAAnomaly(slice, datasetConfig));
+          }
+        } else {
+          datasetLastRefreshTime = dataFrame.getDoubles("timestamp").max().longValue();
+          if (isPartialData(datasetLastRefreshTime, endTime, datasetConfig)) {
+            if (hasMissedSLA(datasetLastRefreshTime, endTime)) {
+              slice = MetricSlice.from(me.getId(), datasetLastRefreshTime + 1, endTime);
+              anomalies.add(createDataSLAAnomaly(slice, datasetConfig));
+            }
+          }
+        }
+      } else if (isPartialData(datasetLastRefreshTime, endTime, datasetConfig)) {
+        // Optimize for the common case - the common case is that the data availability events are arriving
+        // correctly and we need not re-fetch the data to double check.
+        if (hasMissedSLA(datasetLastRefreshTime, endTime)) {
+          anomalies.add(createDataSLAAnomaly(slice, datasetConfig));
+        }
+      }
+    } catch (Exception e) {
+      LOG.error(String.format("Failed to run sla check on metric URN %s", me.getUrn()), e);
+    }
```

Comment:
Why the logic is so complicated?
Can we just compare the data last time vs the expected refresh time?

Reply:
Not quite! We do have that core logic but the complexity comes while dealing with "partial data" and also trying to optimize for the common case at the same time (reduce unnecessary querying of data source).

I have updated this section of code to make it more readable. Please review. Also, this section of code comes from the previous PR (listed here due to refactoring).

Reply:
Thanks. That looks much more clear.